File: gateway/limiter/tokenbucket.go
// Package limiter implements rate-limiting algorithms for the gateway.
//
// Phase 1: in-memory token bucket, one bucket per client key (IP or API key).
// Phase 2 will swap the storage for Redis so multiple gateway replicas share
// state - but the Allow() interface stays the same, which is why it's
// worth getting this abstraction right now.
package limiter

import (
	"sync"
	"time"
)

// bucket holds the state for a single client's token bucket.
type bucket struct {
	tokens     float64
	lastRefill time.Time
}

// TokenBucketLimiter implements the token bucket algorithm:
// each client gets a bucket that holds up to `capacity` tokens.
// Tokens refill continuously at `refillRate` tokens/second.
// Each request costs 1 token. If the bucket is empty, the request is denied.
//
// Why token bucket for v1: it naturally allows short bursts (as long as
// tokens have accumulated) while still enforcing a long-run average rate -
// which is closer to real traffic patterns than a hard fixed-window cap.
type TokenBucketLimiter struct {
	mu         sync.Mutex
	buckets    map[string]*bucket
	capacity   float64
	refillRate float64 // tokens per second
}

// NewTokenBucketLimiter creates a limiter where each client can burst up to
// `capacity` requests, and the bucket refills at `refillRate` tokens/sec.
func NewTokenBucketLimiter(capacity, refillRate float64) *TokenBucketLimiter {
	return &TokenBucketLimiter{
		buckets:    make(map[string]*bucket),
		capacity:   capacity,
		refillRate: refillRate,
	}
}

// Allow reports whether the request identified by `key` (an IP or API key)
// should be permitted right now. It also returns the number of tokens
// remaining after this decision, which we'll surface as a response header.
func (l *TokenBucketLimiter) Allow(key string) (allowed bool, remaining float64) {
	l.mu.Lock()
	defer l.mu.Unlock()

	now := time.Now()
	b, exists := l.buckets[key]
	if !exists {
		// First time we've seen this client: start with a full bucket.
		b = &bucket{tokens: l.capacity, lastRefill: now}
		l.buckets[key] = b
	}

	// Refill based on elapsed time since we last touched this bucket.
	elapsed := now.Sub(b.lastRefill).Seconds()
	b.tokens += elapsed * l.refillRate
	if b.tokens > l.capacity {
		b.tokens = l.capacity
	}
	b.lastRefill = now

	if b.tokens >= 1 {
		b.tokens -= 1
		return true, b.tokens
	}
	return false, b.tokens
}

File: gateway/limiter/tokenbucket.lua
-- Atomic token bucket check-and-consume, run entirely inside Redis.
--
-- Why Lua here: two gateway replicas could otherwise both read "3 tokens
-- left", both decide to allow, and both write "2 tokens left" - silently
-- letting through one more request than the bucket should permit. Redis
-- executes a Lua script as a single atomic operation, so there's no window
-- for another replica's request to interleave with this one.
--
-- Why redis.call('TIME') instead of passing time from Go: every gateway
-- replica has its own system clock, and clock drift between machines would
-- corrupt the refill math. Using Redis's own clock gives every replica a
-- single, consistent source of time.
--
-- KEYS[1] = bucket key, e.g. "ratelimit:ip:203.0.113.7"
-- ARGV[1] = capacity (max tokens)
-- ARGV[2] = refill rate (tokens per second)
-- ARGV[3] = requested tokens for this call (1 per request)
-- ARGV[4] = idle TTL in seconds, so buckets for clients who've stopped
--           making requests eventually expire instead of accumulating in
--           Redis forever.

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local requested = tonumber(ARGV[3])
local ttl = tonumber(ARGV[4])

local time = redis.call('TIME')
local now = tonumber(time[1]) + (tonumber(time[2]) / 1000000)

local bucket = redis.call('HMGET', key, 'tokens', 'timestamp')
local tokens = tonumber(bucket[1])
local timestamp = tonumber(bucket[2])

if tokens == nil then
    -- First time we've seen this key: start with a full bucket.
    tokens = capacity
    timestamp = now
end

local elapsed = math.max(0, now - timestamp)
local filled = math.min(capacity, tokens + (elapsed * refill_rate))

local allowed = 0
local new_tokens = filled
if filled >= requested then
    allowed = 1
    new_tokens = filled - requested
end

redis.call('HMSET', key, 'tokens', new_tokens, 'timestamp', now)
redis.call('EXPIRE', key, ttl)

return { allowed, tostring(new_tokens) }